staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go

```diff
@@ -178,7 +179,11 @@ func (conn *Conn) Open(w http.ResponseWriter, req *http.Request) (string, []io.R
 	go func() {
 		defer runtime.HandleCrash()
 		defer conn.Close()
-		websocket.Server{Handshake: conn.handshake, Handler: conn.handle}.ServeHTTP(w, req)
+		handler := conn.handle
+		if _, ok := conn.protocols[remotecommand.StreamProtocolV5Name]; ok {
```

Comment:
is presence of v5 in conn.protocols correct, or should we be looking at selectedProtocol? I think we should be triggering off of conn.selectedProtocol

Reply:
I also think that it would be clearer what the difference between the protocols is if the reset byte handling is inline and conditional in the existing `handle` function:

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go
index 09f54a49c74..586171f107c 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go
@@ -244,6 +244,8 @@ func (conn *Conn) handle(ws *websocket.Conn) {
 	defer conn.Close()
 	conn.initialize(ws)
 
+	supportsStreamReset := protocolSupportsStreamReset(conn.selectedProtocol)
+
 	for {
 		conn.resetTimeout()
 		var data []byte
@@ -256,6 +258,18 @@ func (conn *Conn) handle(ws *websocket.Conn) {
 		if len(data) == 0 {
 			continue
 		}
+
+		if supportsStreamReset && data[0] == remotecommand.StreamReset {
+			if len(data) != 2 {
+				// error
+			} else {
+				channel := data[1]
+				klog.V(4).Infof("Received half-close signal from client; close %d stream", channel)
+				conn.channels[channel].Close()
+			}
+			continue
+		}
+
 		channel := data[0]
 		if conn.codec == base64Codec {
 			channel = channel - '0'
```

Reply:
where the implementation of `protocolSupportsStreamReset` could either be done positively (only match on v5) or negatively (match on everything other than v1-v4)

Reply:
also, what happens if we get multiple close commands for the same channel? is that specified or tested?

Reply:
consider whether to call this "CLOSE" instead of "RESET" (matches io interface better)

Reply:
* Change signal from RESET to CLOSE
* Removed the V5 handler, conditionally handling CLOSE signal.
* Added error handling around CLOSE signal.
* Added comment that extra CLOSE signals are noops.